proposals/0011-inline-spirv.md

```diff
@@ -0,0 +1,347 @@
+    [execution mode](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Execution_Mode),
+1.  add a new instruction or extended instruction,
+1.  add a new type,
+1.  add a new decoration,
+1.  add a new storage class, or
+1.  add a new builtin.
+
+Most extensions also define
+[capabilities](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Capability)
+that are used to enable the new features, but the capabilities usually do
+nothing on their own.
+
+The proposed solution will add syntax to define each of these features aliased
+to a meaningful name, so that they can be defined in a header file. Then users
+will be able to use the header file, and use the feature by using the name. They
+will not have to see consider the inline SPIR-V. The proposed solution will also
+expect that the SPIR-V referenced is defined in the version of spirv-headers
+that was used to build DXC.
+
+### Execution modes
+
+The existing `vk::ext_execution_mode` is a builtin function. I'm guessing it was
+defined as a builtin function because the original designers want to be able to
+add an execution mode that was not yet defined in spirv-headers. Since we are
+not considering that workflow anymore, I propose that we define a new attribute
+`vk::spvexecutionmode(ExecutionMode)`.
+
+Then suppose we wanted to implement the
+[SPV_KHR_post_depth_coverage](http://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/KHR/SPV_KHR_post_depth_coverage.html)
+extension in a header file.
+
+The header file could be something like
+
+```
+// spv_khr_post_depth_coverage.h
+
+// It would be nice to have this live in a namespace spv::khr::, but not possible with attributes.
+const uint32_t SampleMaskPostDepthCoverageCapabilityId = 4447;
+const uint32_t PostDepthCoverageExecutionModeId = 4446;
+#define SPV_KHR_PostDepthCoverageExecutionMode vk::ext_extension("SPV_KHR_post_depth_coverage"), vk::ext_capability(SampleMaskPostDepthCoverageCapabilityId), vk::spvexecutionmode(spv::khr::PostDepthCoverageExecutionModeId)
+```
+
+Then the user could write:
+
+```
+#include "spv_khr_post_depth_coverage.h"
+
+[[SPV_KHR_PostDepthCoverageExecutionMode]]
+float4 PSMain(...) : SV_TARGET
+{ ...
+}
+```
+
+### Instructions and extended instructions
+
+The existing `vk::ext_instruction` attribute is used to define a function as a
+specific spir-v instruction. Suppose we wanted to implement the
+[SPV_INTEL_subgroups](http://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/INTEL/SPV_INTEL_subgroups.html)
+extension in a header file. It contains 8 new instructions. Each one could be
+defined in the header file as a function, and then the function can be called by
+the users. For example,
+
+```
+[[vk::ext_capability(5568)]]
+[[vk::ext_extension("SPV_INTEL_subgroups")]]
+[[vk::ext_instruction(/* OpSubgroupShuffleINTEL */ 5571)]]
+template<typename T>
+T SubgroupShuffleINTEL(T data, uint32 invocationId);
```

Comment:
Allowing 'template' would be fantastic; as another example, it would allow the implementation of [hitObjectTraceRayNV](https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/NV/SPV_NV_shader_invocation_reorder.html) for Shader-Execution-Reorder in a header.  Currently, if we attempt to do the following, we receive the error "G53006855: an attribute list cannot appear here," and the payload's explicit structure needs to be defined.

```
[[vk::ext_instruction(OpHitObjectTraceRayNV)]]
template<typename T>
void hitObjectTraceRayNV(
    [[vk::ext_reference]] HitObjectNV hitObject,
    RaytracingAccelerationStructure as,
    uint RayFlags,
    uint CullMask,
    uint SBTOffset,
    uint SBTStride,
    uint MissIndex,
    float3 RayOrigin,
    float RayTmin,
    float3 RayDirection,
    float RayTMax,
    [[vk::ext_reference]] [[vk::ext_storage_class(RayPayloadKHR)]] T payload
  );
```